crates/sui-storage/src/write_ahead_log.rs

```diff
@@ -185,32 +210,35 @@ where
         tx: &TransactionDigest,
         cert: &C,
     ) -> SuiResult<Option<DBTxGuard<'a, C>>> {
-        let _mutex_guard = self.mutex_table.acquire_lock(tx).await;
+        let mutex_guard = self.mutex_table.acquire_lock(tx).await;
 
         if self.log.contains_key(tx)? {
-            // We return None instead of a guard, to signal that a tx with this digest is already
-            // in progress.
+            // A concurrent tx will have held the mutex guard until it finished. If the tx is
```

Comment:
Given everything this comment says: should we not then allow another instance of execute for exactly the same transaction to proceed? As in: would it be incorrect to return the DBTxGuard here for this without re-inserting the tx, and allowing the write to progress? Why force a recovery path where a second write might maybe do the job?

Reply:
Hm, there will be issues if this tx includes shared objects. A tx with a cert should already deterministically lead to either a success or abort status, and effects. Otherwise we might lose liveness. So that is a critical bug we should not ignore.

Reply:
agreed. Will add the retry count and be sure to scream loudly if we keep retrying the same tx